board/safety/safety_nissan.h

```diff
@@ -0,0 +1,212 @@
+          controls_allowed = 0;
+        }
+        nissan_brake_prev = brake;
+      }
+    }
+  }
+  return valid;
+}
+
+
+static int nissan_tx_hook(CAN_FIFOMailBox_TypeDef *to_send) {
+  int tx = 1;
+  int addr = GET_ADDR(to_send);
+  int bus = GET_BUS(to_send);
+
+  if (!msg_allowed(addr, bus, NISSAN_TX_MSGS, sizeof(NISSAN_TX_MSGS) / sizeof(NISSAN_TX_MSGS[0]))) {
+    tx = 0;
+  }
+
+  if (relay_malfunction) {
+    tx = 0;
+  }
+
+  // steer cmd checks
+  if (addr == 0x169) {
+    int desired_angle = ((GET_BYTE(to_send, 0) << 10) | (GET_BYTE(to_send, 1) << 2) | ((GET_BYTE(to_send, 2) >> 6) & 0x3));
+    bool lka_active = (GET_BYTE(to_send, 6) >> 4) & 1;
+
+    // offeset 1310 * NISSAN_DEG_TO_CAN
+    desired_angle =  desired_angle - 131000;
+    bool violation = 0;
+
+    if (controls_allowed && lka_active) {
+      // add 1 to not false trigger the violation
+      float delta_angle_float;
+      delta_angle_float = (interpolate(NISSAN_LOOKUP_ANGLE_RATE_UP, nissan_speed) * NISSAN_DEG_TO_CAN) + 1.;
+      int delta_angle_up = (int)(delta_angle_float);
+      delta_angle_float =  (interpolate(NISSAN_LOOKUP_ANGLE_RATE_DOWN, nissan_speed) * NISSAN_DEG_TO_CAN) + 1.;
+      int delta_angle_down = (int)(delta_angle_float);
+      int highest_desired_angle = nissan_desired_angle_last + ((nissan_desired_angle_last > 0) ? delta_angle_up : delta_angle_down);
+      int lowest_desired_angle = nissan_desired_angle_last - ((nissan_desired_angle_last >= 0) ? delta_angle_down : delta_angle_up);
+
+      // Limit maximum steering angle at current speed 
+      int maximum_angle = ((int)interpolate(NISSAN_LOOKUP_MAX_ANGLE, nissan_speed));
+
+      if (highest_desired_angle > (maximum_angle * NISSAN_DEG_TO_CAN)) {
+        highest_desired_angle = (maximum_angle * NISSAN_DEG_TO_CAN);
+      }
+      if (lowest_desired_angle < (-maximum_angle * NISSAN_DEG_TO_CAN)) {
+        lowest_desired_angle = (-maximum_angle * NISSAN_DEG_TO_CAN);
+      }
+
+      // check for violation;
+      violation |= max_limit_check(desired_angle, highest_desired_angle, lowest_desired_angle);
+
+      //nissan_controls_allowed_last = controls_allowed;
+    }
+    nissan_desired_angle_last = desired_angle;
+
+    // desired steer angle should be the same as steer angle measured when controls are off
+    if ((!controls_allowed) &&
+          ((desired_angle < (nissan_angle_meas.min - 1)) ||
+          (desired_angle > (nissan_angle_meas.max + 1)))) {
+      violation = 1;
+    }
+
+    // no lka_enabled bit if controls not allowed
+    if (!controls_allowed && lka_active) {
+      violation = 1;
+    }
+
+    if (violation) {
+      controls_allowed = 0;
+      tx = 0;
+    }
+  }
+  return tx;
+}
+
+
+static int nissan_fwd_hook(int bus_num, CAN_FIFOMailBox_TypeDef *to_fwd) {
+  int bus_fwd = -1;
+  int addr = GET_ADDR(to_fwd);
+
+  if (bus_num == 0) {
+    // 0x20b is CruiseThrottle
+    int block_msg = (addr == 0x20b);
```

Comment:
why do you need to block the cruise throttle from bus 0 to 2?

Reply:
This is probably a bad message name. This message contains cruise control button status and gas pedal position.

This message is generated on the EON, it basically passes through all button presses unless OP wants to cancel ACC/LKAS then it will set the cancel button signal to true.

Reply:
as discussed, instead of blocking the message, it's better to just spam this message and put a safety tx check that the only allowed button is "cancel"